# Part 2: Data cleaning

    > We have cleaned the files now we have to treat the missing values we found in rest of the columns.
    > This time we will import from the cleaned csv files

In [1]:
import numpy as np
import pandas as pd

import warnings
warnings.filterwarnings('ignore')

In [2]:
#creating paths

rounds_path = '/Users/manish/Documents/Projects/data_science/investment_strategy/data/clean_data/rounds2_clean.csv' 
companies_path = '/Users/manish/Documents/Projects/data_science/investment_strategy/data/clean_data/companies_clean.csv'


In [3]:
#creating dataframes
rounds = pd.read_csv(rounds_path)
companies = pd.read_csv(companies_path)


In [4]:
#verifying the permalink columns unique entries in both the columns

print('rounds', len(rounds.company_permalink.unique()))
print('companies', len(companies.permalink.unique()))

rounds 66368
companies 66368


    > It seems to be working correctly.
    > Now our job is missing value treatments 

# Missing value treatment

In [5]:
#checking missing values in bothe dataframes

print(rounds.isnull().sum())
print('--'*20)
print(companies.isnull().sum())

company_permalink              0
funding_round_permalink        0
funding_round_type             0
funding_round_code         83809
funded_at                      0
raised_amount_usd          19990
dtype: int64
----------------------------------------
permalink            0
name                 1
homepage_url      5058
category_list     3148
status               0
country_code      6958
state_code        8547
region            8030
city              8028
founded_at       15221
dtype: int64


    > lets create a master data frame for ease of analysis
    > we can use pd.merge to merge on company_permalink column
    > after merging we can drop any one of the permalink column as they are redundant

In [6]:
#merging the two files
master = pd.merge(companies, rounds, how = 'inner',
                  left_on = 'permalink', right_on = 'company_permalink')
master = master.drop(columns = ['company_permalink'])
master.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2               /organization/-qounter                :Qounter   
3  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
4                /organization/0-6-com                 0-6.com   

             homepage_url                                      category_list  \
0      http://livfame.com                                              Media   
1  http://www.qounter.com  Application Platforms|Real Time|Social Network...   
2  http://www.qounter.com  Application Platforms|Real Time|Social Network...   
3     http://oneofthem.jp                                  Apps|Games|Mobile   
4      http://www.0-6.com                                        Curated Web   

      status country_code state_code      region           city  founded_at  \
0  operating          IND         16      Mumbai         Mumbai         NaN   
1  operating          USA         DE  DE - Other  Delaware City  04-09-2014   
2  operating          USA         DE  DE - Other  Delaware City  04-09-2014   
3  operating          NaN        NaN         NaN            NaN         NaN   
4  operating          CHN         22     Beijing        Beijing  01-01-2007   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

In [7]:
master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 114949 entries, 0 to 114948
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   permalink                114949 non-null  object 
 1   name                     114948 non-null  object 
 2   homepage_url             108815 non-null  object 
 3   category_list            111539 non-null  object 
 4   status                   114949 non-null  object 
 5   country_code             106271 non-null  object 
 6   state_code               104003 non-null  object 
 7   region                   104782 non-null  object 
 8   city                     104785 non-null  object 
 9   founded_at               94428 non-null   object 
 10  funding_round_permalink  114949 non-null  object 
 11  funding_round_type       114949 non-null  object 
 12  funding_round_code       31140 non-null   object 
 13  funded_at                114949 non-null  object 
 14  rais

    > there we go we have succesfully merged the dataframes
    > Also the company_permalink column has been dropped

In [8]:
#missing entries in master df
print(master.isnull().sum())

permalink                      0
name                           1
homepage_url                6134
category_list               3410
status                         0
country_code                8678
state_code                 10946
region                     10167
city                       10164
founded_at                 20521
funding_round_permalink        0
funding_round_type             0
funding_round_code         83809
funded_at                      0
raised_amount_usd          19990
dtype: int64


In [9]:
#percent of missing in each columns
print(100*master.isnull().sum()/len(master.index))

permalink                   0.000000
name                        0.000870
homepage_url                5.336280
category_list               2.966533
status                      0.000000
country_code                7.549435
state_code                  9.522484
region                      8.844792
city                        8.842182
founded_at                 17.852265
funding_round_permalink     0.000000
funding_round_type          0.000000
funding_round_code         72.909725
funded_at                   0.000000
raised_amount_usd          17.390321
dtype: float64


Clearly, the column ```funding_round_code``` is useless (with about 73% missing values). Also, for the business objectives given, the columns ```homepage_url```, ```founded_at```, ```state_code```, ```region``` and ```city``` need not be used.

In [10]:
#dropping the unnecessary columns
master = master.drop(columns = ['funding_round_code', 'homepage_url',
                                'founded_at', 'state_code', 'region', 
                                'city'])
master.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2               /organization/-qounter                :Qounter   
3  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
4                /organization/0-6-com                 0-6.com   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2  Application Platforms|Real Time|Social Network...  operating          USA   
3                                  Apps|Games|Mobile  operating          NaN   
4                                        Curated Web  operating          CHN   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

    funded_at  raised_amount_usd  
0  05-01-2015         10000000.0  
1  14-10-2014                NaN  
2  01-03-2014           700000.0  
3  30-01-2014          3406878.0  
4  19-03-2008          2000000.0

In [11]:
#percent of missing in each columns
print(round(100*master.isnull().sum()/len(master.index),2))

permalink                   0.00
name                        0.00
category_list               2.97
status                      0.00
country_code                7.55
funding_round_permalink     0.00
funding_round_type          0.00
funded_at                   0.00
raised_amount_usd          17.39
dtype: float64


Missing columns include ```category_list```, ```country_code``` and ```raised_amount_usd```. 

We can not simply delete these columns as ```category_list``` will be used for merging with the mapping file.

```country_code``` and ```raised_amount_usd``` are useful from business perspective.

We have to carefully tread through the ```raised_amount_usd``` column as it has about ```17%``` missing values



In [12]:
#some stats of the raised amount column
master.raised_amount_usd.describe()

count    9.495900e+04
mean     1.042687e+07
std      1.148212e+08
min      0.000000e+00
25%      3.225000e+05
50%      1.680511e+06
75%      7.000000e+06
max      2.127194e+10
Name: raised_amount_usd, dtype: float64

The `mean` amount of funding is `10 million USD`.
The `median` is about `1.7 million USD`.
The highest amount invested is about `21.7 billion USD`
The data is highly skewed and has very large outliers. This clearly inflate the mean.

This suggests we have no other option but to delete the missing values in `raised_amount_usd`
as we can not impute them with mean or median

In [13]:
#dropping the missing values 
master = master[~np.isnan(master.raised_amount_usd)]

In [14]:
print(master.isnull().sum())
print('--'*20)
print(round(100*master.isnull().sum()/len(master.index),2))

permalink                     0
name                          1
category_list              1044
status                        0
country_code               5851
funding_round_permalink       0
funding_round_type            0
funded_at                     0
raised_amount_usd             0
dtype: int64
----------------------------------------
permalink                  0.00
name                       0.00
category_list              1.10
status                     0.00
country_code               6.16
funding_round_permalink    0.00
funding_round_type         0.00
funded_at                  0.00
raised_amount_usd          0.00
dtype: float64


In [15]:
master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 94959 entries, 0 to 114948
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   permalink                94959 non-null  object 
 1   name                     94958 non-null  object 
 2   category_list            93915 non-null  object 
 3   status                   94959 non-null  object 
 4   country_code             89108 non-null  object 
 5   funding_round_permalink  94959 non-null  object 
 6   funding_round_type       94959 non-null  object 
 7   funded_at                94959 non-null  object 
 8   raised_amount_usd        94959 non-null  float64
dtypes: float64(1), object(8)
memory usage: 7.2+ MB


We can see the only `country_code` and `category_list` have missing values left 
along with one entry in the `name` column. 

Lets check the country_code

In [16]:
round(100*master.country_code.value_counts()/master.country_code.count(), 2)

USA    69.63
GBR     5.63
CAN     2.94
CHN     2.16
IND     1.85
       ...  
HND     0.00
GGY     0.00
TGO     0.00
MNE     0.00
SEN     0.00
Name: country_code, Length: 134, dtype: float64

It looks like `USA` is the country with most companies around ``69.63%`` 
of the non null values.Followed by `GBR`, `CAN`, `CHN` and `IND` with 
`5.63%`, `2.94%`, `2.16%` and `1.85%` of the total countries.

Either we can delete the missing values or impute them with USA. 
But since the % missing values in country_code is `~6%` we can del them as we have 
enough data to carry on the analysis

Also since the missing category_list is about `0.65%` we can drop those entries as well.

In [17]:
#deleting the entries with missing country_code
master = master[~pd.isnull(master.country_code)]
master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 89108 entries, 0 to 114947
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   permalink                89108 non-null  object 
 1   name                     89107 non-null  object 
 2   category_list            88529 non-null  object 
 3   status                   89108 non-null  object 
 4   country_code             89108 non-null  object 
 5   funding_round_permalink  89108 non-null  object 
 6   funding_round_type       89108 non-null  object 
 7   funded_at                89108 non-null  object 
 8   raised_amount_usd        89108 non-null  float64
dtypes: float64(1), object(8)
memory usage: 6.8+ MB


In [18]:
#printing the misising value after deleting the missing country_code entries
print(round(100*master.isnull().sum()/len(master.index),2))

permalink                  0.00
name                       0.00
category_list              0.65
status                     0.00
country_code               0.00
funding_round_permalink    0.00
funding_round_type         0.00
funded_at                  0.00
raised_amount_usd          0.00
dtype: float64


In [19]:
#deleting rows with missing category_list
master = master[~pd.isnull(master.category_list)]
master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 88529 entries, 0 to 114947
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   permalink                88529 non-null  object 
 1   name                     88528 non-null  object 
 2   category_list            88529 non-null  object 
 3   status                   88529 non-null  object 
 4   country_code             88529 non-null  object 
 5   funding_round_permalink  88529 non-null  object 
 6   funding_round_type       88529 non-null  object 
 7   funded_at                88529 non-null  object 
 8   raised_amount_usd        88529 non-null  float64
dtypes: float64(1), object(8)
memory usage: 6.8+ MB


In [20]:
#printing the misising value after deleting the missing category_list entries
print(round(100*master.isnull().sum()/len(master.index),2))

permalink                  0.0
name                       0.0
category_list              0.0
status                     0.0
country_code               0.0
funding_round_permalink    0.0
funding_round_type         0.0
funded_at                  0.0
raised_amount_usd          0.0
dtype: float64


In [21]:
#we can see only one entry without name. 
#lets have a look
master.loc[master.name.isnull(),:]

permalink name category_list  status country_code  \
98692  /organization/tell-it-in  NaN      Startups  closed          USA   

                               funding_round_permalink funding_round_type  \
98692  /funding-round/9c987e616755a78c51a4aa67c27a2a93               seed   

        funded_at  raised_amount_usd  
98692  01-03-2012            25000.0

In [22]:
#changing the name at the above loc to Tell_it_in
master.loc[98692,'name'] = 'Tell_it_in'

In [23]:
#checking if it worked
master.loc[master.name.isnull(),:]

Empty DataFrame
Columns: [permalink, name, category_list, status, country_code, funding_round_permalink, funding_round_type, funded_at, raised_amount_usd]
Index: []

In [24]:
#checking the name at the changed location
master.loc[98692,'name']

'Tell_it_in'

In [25]:
#checking for missing values
print(master.isnull().sum())

permalink                  0
name                       0
category_list              0
status                     0
country_code               0
funding_round_permalink    0
funding_round_type         0
funded_at                  0
raised_amount_usd          0
dtype: int64


In [26]:
#checking the cleaned data
master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 88529 entries, 0 to 114947
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   permalink                88529 non-null  object 
 1   name                     88529 non-null  object 
 2   category_list            88529 non-null  object 
 3   status                   88529 non-null  object 
 4   country_code             88529 non-null  object 
 5   funding_round_permalink  88529 non-null  object 
 6   funding_round_type       88529 non-null  object 
 7   funded_at                88529 non-null  object 
 8   raised_amount_usd        88529 non-null  float64
dtypes: float64(1), object(8)
memory usage: 8.8+ MB


We have treated all the missing values.
Now we have `88529` out of `114948` entries left after clean-up.
We have about `78%` of our initial data. Which is low but as the data has `~89K` 
entries, we can do some solid analysis to them.

Now we can put the cleaned master data to a csv file 

In [27]:
#creating clean master csv file
#master.to_csv("master_clean.csv", index = False)

### Lets move to the Analysis notebook